Added support for multiple transforms in the same SVG element

ksachdeva opened the pull request:
## Motivation
At present, the svg-engine can parse only one transform value from the attribute value.

## Overview / Explanation for Changes

Below mentioned transform attr values are valid and are generated by various plotting libraries

```
<g transform="translate(70.651477 322.182437)scale(0.1 -0.1)">
<g transform="translate(70.651477 322.182437), scale(0.1 -0.1)">
<g transform="translate(70.651477 322.182437) scale(0.1 -0.1)">
```

At present, the transform attr parsing logic is assuming that there is only 1 transform for e.g. either translate or scale which makes manim crash later.

This changeset is fixing that issue and at the same time adding the support to handle multiple transform.

The code to parse the transform attr value has been inspired from 
https://github.com/cjlano/svg/blob/master/svg/svg.py#L75

That said, I have added the support in such a way that once the parsing is done the rest of the original operations are kept the same to not introduce a regression.

## Testing Status
- Ran the test suite to arrest for any regression introduced by this change in parsing
- Ran manual verification tests on few svgs of mine
- Tried to add a test but it is looking for some "npz" file and at present I do not know how to add one


## Acknowledgements
- [X] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [X] I have chosen a descriptive PR title (see top of PR template for examples)


<!-- Do not modify the lines below. -->
## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The PR title is descriptive enough

MrMallIronmaker reviewed (changes requested):
Code looks good! I have two minor changes noted. 

Thanks for giving a shot at implementing tests - I think it'll be good to add them before merging. Take a look at https://github.com/ManimCommunity/manim/wiki/Testing. The key is that generating test data is done with a call to `set_test_scene`.

MrMallIronmaker commented:
One more thing - the translate and two-paramater scale have tests, but matrix and one-parameter scale don't yet. (https://codecov.io/gh/ManimCommunity/manim/src/master/manim/mobject/svg/svg_mobject.py - if you 404, try logging in)

If you have svg files to test already, it would be doubly helpful to add those tests too.

ksachdeva commented on manim/mobject/svg/svg_mobject.py:
```diff
@@ -432,39 +432,62 @@ def handle_transforms(self, element, mobject):
+        # [^)]*    == anything but a closing parenthese
+        # '|'.join == OR-list of SVG transformations
+        transforms = re.findall(
+            "|".join([x + r"[^)]*\)" for x in transform_names]), transform_attr_value
```
agree; changing it

ksachdeva commented on manim/mobject/svg/svg_mobject.py:
```diff
@@ -432,39 +432,62 @@ def handle_transforms(self, element, mobject):
+                x, y = t[1]
+                mobject.shift(x * RIGHT + y * DOWN)
+
+            # TODO: handle rotate, skewX and skewY
```
for now, changing the doc but leaving the comment so as to remember to handle it

MrMallIronmaker replied:
Do you plan to handle this before merging? If not, can you change the op_name tests to elifs and at the very end give a warning that op_name is not handled and the svg parser is ignoring it. 

[See e.g. VMobjectFromSVGPath for example warning code]

ksachdeva replied:
this is done now.

ksachdeva commented:
@MrMallIronmaker I have incorporated your suggestion on naming and some more. Note following -

- Loops are merged (had two of them earlier for debugging purposes)
- Some more appropriate naming

Earlier I did not want to touch your original code but now in the "matrix" handling, I have some renaming. Would appreciate if you can review that I have not made any silly mistake as working with python is like playing with daggers :)

Have added the test SVG file (thanks for your pointer to the wiki).

Also, the scaling is not working properly; you are more knowledgeable on that aspect so would be good if you can see what is a remedy for that. Use my SVG to see that the coordinates in this plot are not placed at appropriate position. I did mess with scaling but could not get to anywhere.

Thanks.

MrMallIronmaker commented on tests/test_graphical_units/test_img_and_svg.py:
```diff
@@ -169,6 +169,13 @@ def construct(self):
         self.wait()
 
 
+class MutlipleTransformTest(Scene):
```
(nitpick) Multiple is misspelled here

ksachdeva replied:
:( .... have resolved it

MrMallIronmaker commented on tests/test_graphical_units/img_svg_resources/multiple_transforms.svg:
```diff
@@ -0,0 +1,636 @@
+<?xml version="1.0" encoding="utf-8" standalone="no"?>
```
I presume this test's result was visually inspected for correctness?

ksachdeva replied:
correct

MrMallIronmaker commented on manim/mobject/svg/svg_mobject.py:
```diff
@@ -432,39 +432,55 @@ def handle_transforms(self, element, mobject):
+        transform_regex = "|".join([x + r"[^)]*\)" for x in transform_names])
+        transforms = re.findall(transform_regex, transform_attr_value)
+
+        number_regex = r"[-+]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][-+]?\d+)?"
```
Not a blocking issue, but I wonder if this is similar to the `string_to_numbers` call

ksachdeva replied:
for now, I would prefer to use the regex as I have taken it from the mentioned repo and have tested multiple SVGs through it (from `transforms` point of view). Maybe a follow-up to see if string_to_numbers is okay and/or modify the string_to_numbers to use this regex.

MrMallIronmaker reviewed (changes requested):
Hopefully the test cases make python less dagger-ful. Let me know if you have specific questions you'd like review on; otherwise the changes you've made are good enough to pass a quick glance.

One thing that Manim does is that it automatically scales up/down any SVG to a reasonable size. I presume this came from the fact that Grant didn't want to mess with all the measurement conventions that are part of the SVG spec.

However, that means scaling at the top level is essentially ignored. I think it'd work if one sibling <g> or <path> is scaled but not the other.

Does the scaling work on master?

ksachdeva commented:
@MrMallIronmaker I have taken into account all your feedback except for number_regx vs string_to_number. Also, I do not intend to support rotate, skewX, skewY but have added warning messages for not handled transform and have also added test SVG files using content from here - https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/transform.

Have visually verified that the warning messages are displayed/generated.

MrMallIronmaker commented:
I appreciate the work done in by adding the new tests. Unfortunately, I wasn't clear. I had meant to ask to add tests for the matrix transform and single-argument scale. While we had the functionality written, we didn't have the tests to test them. Because the skew and rotate functionalities aren't added yet, I don't think they should be tested.

Do you want to implement skew and rotate? If so, would you like to do it in this branch or in a separate branch? If not, you can remove the skew and rotate tests.

Also, did you figure out the scaling issues? If not, can you describe the issues in more detail and / or tell me how to replicate the problem?

ksachdeva commented:
> I appreciate the work done in by adding the new tests. Unfortunately, I wasn't clear. I had meant to ask to add tests for the matrix transform and single-argument scale. While we had the functionality written, we didn't have the tests to test them. 

Ah, it seems I misunderstood. Apologies for that. I will try to figure them out.

Question - So far I have not figured out an automated way of testing. When u say add tests do you mean to add some more SVGs and then visually verify them?

> Because the skew and rotate functionalities aren't added yet, I don't think they should be tested.
> 
> Do you want to implement skew and rotate? If so, would you like to do it in this branch or in a separate branch? If not, you can remove the skew and rotate tests.

I will not be implementing the skew and rotate functionality at the moment. Not in this PR at least. Interestingly the purpose of this PR was different than what it is now.

My initial objective was to simply use the SVG that has multiple transforms go through the manim without crashing. In other words, a change to the parsing logic. 

Ideally, a bug fix, should not add new functionality but then this new way of parsing ended up creating the support for handling multiple transforms as a side effect. Then one of the contributors changed the title of the PR to indicate that it is an enhancement and not a bug fix. 

I considered it to be a bug fix and hence was not keen on touching your logic of matrix, scale, and translate and by extension to this not implementing the additional features of rotate and skew. 

Regardings tests, there are two reasons why it may be interesting to keep them

a) Helps in testing that we are indeed going in the warning path and generating code coverage
b) Be ready when you, me, or someone will add the support for rotate and skew. As such it is not a bad idea to have tests anytime. TDD (purist) will even argue to write tests first ...(but I am not arguing :) )

> 
> Also, did you figure out the scaling issues? If not, can you describe the issues in more detail and / or tell me how to replicate the problem?

I will create a separate issue for that with the generated image so that it is easier for you to replicate the problem.

ksachdeva commented:
@MrMallIronmaker I have created the issue here https://github.com/ManimCommunity/manim/issues/1171

Please note that to reproduce the scale issue you need to have the changes from this pull request incorporated else you would see a crash in 0.4 or master because of multiple transforms in the SVG.

ksachdeva commented:
@MrMallIronmaker I have added the tests for matrix, scale, and translate as well. These SVGs have both the combinations i.e. 1 and 2 arguments and then some.

Have verified the stuff visually.

Interestingly the bug caught is in the handling of translate transform. Apparently, it can also take 1 argument. I have fixed this aspect as not handling it leads to an exception.

MrMallIronmaker reviewed (approved):
Well put, I think you've made your case. I'll flag a separate the issue regarding rotate and skew for another time. Thanks for the work and the responsiveness!

ksachdeva commented:
> Well put, I think you've made your case. I'll flag a separate the issue regarding rotate and skew for another time. Thanks for the work and the responsiveness!

Thanks, @MrMallIronmaker for the review and time. Much appreciated.
